Template Parts: allow selection in page content focus mode

noisysocks opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Allows Template Part blocks (Header and Footer) to be selected when editing a page in the site editor in page content focus mode.

## Why?
See https://github.com/WordPress/gutenberg/issues/59970#issuecomment-2005483531.

## How?
- Set the editing mode of `core/template-part` blocks to `contentOnly`, so that you can select it.
- Set the editing mode of children of `core/template-part` blocks to `disabled`, so that you're forced to click _Edit_ in the toolbar.
- Changes `BlockToolbar` to show the block icon / block name on blocks that have a `contentOnly` editing mode, so that you can see the name of the template part when you've selected it.

## Testing Instructions
1. Go to Appearance → Editor → Pages and create or edit page.
2. Click on the header or footer.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/612155/6d9f9c36-e45b-4f6f-befc-5869b8bb2976

noisysocks committed:
Allow selecting template parts in page content focus mode

noisysocks committed:
Show block name/icon in toolbar when editing mode is contentOnly

noisysocks committed:
Only show page content blocks in PageContent sidebar section

SaxonF reviewed (commented):
Working well in testing. Nice work! I think its an improvement and finds a better balance whilst still locking templates which aren't modified as much as site navigation. @richtabor you happy to start here?

andrewserong commented on packages/block-editor/src/components/block-switcher/index.js:
```diff
@@ -229,9 +229,8 @@ export const BlockSwitcher = ( { clientIds } ) => {
-	const hideDropdown = ! hasBlockStyles && ! canRemove;
 
-	if ( hideDropdown ) {
+	if ( disabled || ( ! hasBlockStyles && ! canRemove ) ) {
```
Tiny nit: Is it worth adding a simple comment before this line? Previously the `hideDropdown` const communicated that this block was about outputting the components without the dropdown, so it might help readability slightly to have a comment in its place?

andrewserong reviewed (commented):
Nice, this feels pretty good to use to me @noisysocks, so feels like a good place to start IMO! A couple of minor details:

The test `create a new page, edit template and toggle page template preview ` appears to be failing expecting the `Edit your template to edit this block` notice, so some of the e2es might need updating.

While this is testing nicely for me in the site editor's view of editing a page, in the post editor with template preview mode switched on, it's possible to select the header or footer, and they're visible in the list view, but there's no option to Edit, so it isn't really possible to do anything. In this case, if the blocks cannot be edited in the current context, what would the desired behaviour be?

https://github.com/WordPress/gutenberg/assets/14988353/c3022c79-94f6-496d-8603-7efb6e8bcc61

jasmussen commented:
Nice work. When you're editing a page, this seems like a good flow. To clarify a hidden nuance here, which is related to editing of inline template parts as [also discussed here](https://github.com/WordPress/gutenberg/pull/59883#issuecomment-2005006973): in this branch inline editing of template parts is still possible when you select a _template_ in the site editor. So the main change here is for _pages only_, where now you can select the TP, but only edit it in isolation.

Since we are already having a special case for pages, this seems okay to me, but would love also a gut-check from @richtabor.

jameskoster commented:
Nice, this is working well for me and significantly shortens the journey to go and edit template parts from a page 🚀  

A couple of small quirks we'll likely want to address:

* Hover / focus outline should be purple rather than theme-color. 
* When the page loads, template parts show a dotted outline which fades away – I assume due to `contentOnly`? I'm not sure we need to specifically draw attention to the edit-ability of template parts?
* The appending inserter shows up when you select a TP, which allows partial editing outside of focus mode:

<img width="1456" alt="Screenshot 2024-03-20 at 10 00 50" src="https://github.com/WordPress/gutenberg/assets/846565/6f7a78b1-106f-4a36-be9b-48da6875aa57">

* This only happens with my footer for some reason (stock `page.html` template in TT4), but when I select then focus away there's a double-border effect, one purple and one blue:

<img width="679" alt="Screenshot 2024-03-20 at 10 13 38" src="https://github.com/WordPress/gutenberg/assets/846565/dbcc651e-c99a-455e-82ad-11abf4437d20">

richtabor commented:
> So the main change here is for pages only, where now you can select the TP, but only edit it in isolation.

That's my understanding as well, and on point with my expectations of this effort. 

> Hover / focus outline should be purple rather than theme-color.
> When the page loads, template parts show a dotted outline which fades away – I assume due to contentOnly? I'm not sure > we need to specifically draw attention to the edit-ability of template parts?

I think the whole dotted lines fading in work needs revisiting. It's not quite landing right.

richtabor reviewed (approved):
This is certainly a step forward. 

I'd like to see further explorations on an inline focus/isolated view, rather than navigating to the disconnected view next.

noisysocks committed:
Fix pages.spec.js

noisysocks committed:
Improve clarify of if()

noisysocks committed:
Move template part Edit button from @wordpress/edit-site to @wordpress/editor

noisysocks commented:
> The test create a new page, edit template and toggle page template preview  appears to be failing expecting the Edit your template to edit this block notice, so some of the e2es might need updating.

Fixed 👍 

> While this is testing nicely for me in the site editor's view of editing a page, in the post editor with template preview mode switched on, it's possible to select the header or footer, and they're visible in the list view, but there's no option to Edit, so it isn't really possible to do anything. In this case, if the blocks cannot be edited in the current context, what would the desired behaviour be?

Great catch @andrewserong. My read on the fix for this is that we should be reducing the number of differences between the site editor and the post editor. I know @youknowriad had been doing a lot in this area. So with that in mind in 34e9348a554a190f4ce888003bd9a711fd7a3c29 I moved the code that adds the _Edit_ button to the Template Part toolbar from `@wordpress/edit-site` to `@wordpress/editor` so that it appears in both editors.

ramonjd commented:
Nice stuff.

I tested this in the site and post editors. Works well with theme template parts and custom template parts.

Highlight color appears, for me, correctly on mouseover and on touch. I couldn't break it after the latest updates.

https://github.com/WordPress/gutenberg/assets/6458278/e314ddfb-1418-472e-a315-63a5f2400ebf

🚢

andrewserong commented:
> So with that in mind in https://github.com/WordPress/gutenberg/commit/34e9348a554a190f4ce888003bd9a711fd7a3c29 I moved the code that adds the Edit button to the Template Part toolbar from @wordpress/edit-site to @wordpress/editor so that it appears in both editors.

Thanks for following up. Unfortunately, it doesn't appear to be working properly for me in the post editor. When I go to click Edit, the template part appears to be in content only mode. Then if I click to switch to edit the template, I get a blank screen:

https://github.com/WordPress/gutenberg/assets/14988353/e8088946-4153-40e7-b602-6c2999ef7661

Not sure what's going on there, but if it's too tricky to get working, would it work to only expose the template parts in the site editor? Or does that introduce more complexity again 🤔

ramonjd commented:
> When I go to click Edit, the template part appears to be in content only mode. Then if I click to switch to edit the template, I get a blank screen

I'm noticing this too, but also in trunk

https://github.com/WordPress/gutenberg/assets/6458278/4b484c0d-88df-47e9-a3d3-97baa1db4ab4

noisysocks commented:
Technically it's an existing issue (you also run into it if you select _Template preview_ and then _Edit template_) but I'll see if it's easy to include a fix here.

noisysocks committed:
Rendering mode should be 'all' in post editor when editing a template or template part

noisysocks commented:
> Hover / focus outline should be purple rather than theme-color.
> This only happens with my footer for some reason (stock page.html template in TT4), but when I select then focus away there's a double-border effect, one purple and one blue:

I'm not seeing these issues. Maybe I'm missing something?

> The appending inserter shows up when you select a TP, which allows partial editing outside of focus mode:

What's a TP sorry? 😄 

> When the page loads, template parts show a dotted outline which fades away – I assume due to contentOnly? I'm not sure we need to specifically draw attention to the edit-ability of template parts?

Yeah that's right. I agree it's not worth highlighting the template part. I struggled to come up with a good solution for this because the flashing needs to be implemented in an agnostic way that works across all locked blocks (e.g. patterns with `templateLock=contentOnly`). I'll dig a little more.

andrewserong commented:
> but I'll see if it's easy to include a fix here. edit: done.

That's testing nicely for me, now! 🚀

youknowriad commented on packages/edit-post/src/editor.js:
```diff
@@ -65,6 +65,11 @@ function Editor( {
+				defaultRenderingMode:
+					currentPost.postType === 'wp_template' ||
+					currentPost.postType === 'wp_template_part'
+						? 'all'
```
This mode has been removed in https://github.com/WordPress/gutenberg/pull/58935 are we restoring it back? Why?

noisysocks replied:
Aha my mistake. I want `post-only`. The tricky part is setting or resetting the editing mode when you navigate forward or backwards.

noisysocks replied:
Fixed in 120c78555b8560bba67d3db5349e10a39916233f.

Also updated some docs missed in https://github.com/WordPress/gutenberg/pull/58935: c54239624a160cb4881012881775cacef095bbb3.

youknowriad commented on packages/editor/src/components/provider/disable-non-page-content-blocks.js:
```diff
@@ -5,49 +5,68 @@ import { useSelect, useDispatch } from '@wordpress/data';
+ * Component that when rendered, makes it so that the site editor allows only
+ * page content to be edited.
+ */
+export default function DisableNonPageContentBlocks() {
```
Not specific to this PR but I see this component as a hack. Also the `setBlockEditingMode` and `useBlockEditingMode` APIs have proven to be problematic in a lot of ways. Multiple conflicts and race conditions, I'd love to absorb these in the block editor framework/package instead (maybe through a setting or something) but we'd need to think properly about the API that we'd want to provide there.

noisysocks replied:
Yeah agreed and I keep running into limitations too. For example in this PR I want to disable insertion but not selection on the Template Part and disable all of its children. `contentOnly` doesn't really cut it for the former. There were similar issues with synced patterns.

We also have lots of ways to do similar things. I feel that with the right abstraction we could unify `tempateLock`, the `lock` attribute, and `setBlockEditingMode`.

youknowriad replied:
> We also have lots of ways to do similar things. I feel that with the right abstraction we could unify tempateLock, the lock attribute, and setBlockEditingMode.

I prefer templateLock instead because it's declarative while setBlockEditingMode is more subject to race conditions. I think the framework should be able to "guess" the mode, rather the consumer providing it for each block.

youknowriad commented on packages/editor/src/hooks/template-part-edit.js:
```diff
@@ -84,6 +73,6 @@ export const withEditBlockControls = createHigherOrderComponent(
-	'core/edit-site/template-part-edit-button',
+	'core/editor/template-part-edit-button',
 	withEditBlockControls
 );
```
What's (which dependency) preventing us from bundling this code within the template part block code directly?

noisysocks replied:
Nothing stands out to me. I'll try it.

noisysocks replied:
👍 27962d05c5

jameskoster commented:
> I'm not seeing these issues. Maybe I'm missing something?

At 5s in the video in the OP you can see the outline that appears when you hover the template part is blue, but when you select it it becomes purple. I think it should be purple on hover too. 

Similarly the outline is blue when the template part is selected but not focussed. Basically the outlines should always be purple.

bacoords commented:
I just want to add that conceptually I'm a fan of this approach. I did not engage in extensive, testing, just brief. Small issues/questions I noted:

- Is there a way to still respect root padding when editing a root-level template part like header/footer? I assume there's a technical challenge there.
- Is it possible to include the "double click" to edit a template part that (I believe) is currently available?
- How can we edit the post title on pages in the site editor if the template doesn't include a post title - is putting a title block in the content the best approach? How does this effect pages where the title is not actually visible on the page content?

noisysocks committed:
Revert "Rendering mode should be 'all' in post editor when editing a template or template part"

This reverts commit 348889e717343d1e5c988ba586955d1a26252d24.

noisysocks committed:
Reset and remember rendering mode when navigating in post editor

noisysocks committed:
Remove unsupported rendering mode from docs

noisysocks committed:
Add 'Edit' button to Template Part block natively

noisysocks committed:
Disable appending into a template part

noisysocks commented:
> * The appending inserter shows up when you select a TP, which allows partial editing outside of focus mode:

I fixed this in 19516d7588 by disabling appending when a template part is set to `contentOnly` which is not a great solution. I think we need to iterate on `setBlockEditingMode` to accomodate cases like this more, but this will do for now.

noisysocks committed:
Always show purple outline on template parts

noisysocks commented:
> Similarly the outline is blue when the template part is selected but not focussed. Basically the outlines should always be purple.

OK think I've fixed this. Thanks @jameskoster. Feel free to push more CSS tweaks if I missed something.

noisysocks commented:
> * Is there a way to still respect root padding when editing a root-level template part like header/footer? I assume there's a technical challenge there.

What do you mean by root padding?

> * Is it possible to include the "double click" to edit a template part that (I believe) is currently available?

> * I don't think we currently have a double click shortcut for template parts? Or at least locally in `trunk` when I double click a template part nothing happens beyond the inner block that I clicked on being selected which I'd expect to happen in any block.

Good question. Right now one would have to switch to the post editor if there's no Post Title block. It could be worth exploring allowing the title to be set in one of the sidebars (black W sidebar on left, or page sidebar on right). It's beyond the scope of this PR though.

noisysocks commented:
Thanks everyone! I've addressed / responded to all feedback. I'll merge once there's an approving code review.

bacoords commented:
> What do you mean by root padding?

I'll open a separate issue for this.

> I don't think we currently have a double click shortcut for template parts? Or at least locally in trunk when I double click a template part nothing happens beyond the inner block that I clicked on being selected which I'd expect to happen in any block.

Currently if you click the surrounding template or a template part from a page in the site editor you'll see a toast notification, but on a double click you get the modal that allows you to enter: 

<img width="418" alt="Screenshot 2024-03-21 at 8 29 06 PM" src="https://github.com/WordPress/gutenberg/assets/6867360/26ee3a60-52d8-40a0-8cf7-eaa31b1cbf52">

The caveat being (as I think you implied) that it allows you to edit the "template" whether you've double clicked over a template or a template part. But in trunk once in the template you are at that point able to edit either one in the same context. 

In this PR you can see the same thing happen when you double click to edit the template around the page content, but not when you double click to edit the template part. That region previously responded to a double click but doesn't any longer. 

Just noting that the behavior is inconsistent and it would be great to be able to say definitely to our users "you can double click to enter into editing global content". Understand if that requires a new issue/PR.

andrewserong commented on packages/edit-post/src/hooks/use-navigate-to-entity-record.js:
```diff
@@ -34,32 +38,41 @@ export default function useNavigateToEntityRecord(
+		]
 	);
 
-	const initialPost = useMemo( () => {
-		return {
-			type: initialPostType,
-			id: initialPostId,
-		};
-	}, [ initialPostType, initialPostId ] );
```
Why is `initialPost` removed from here and moved further down into the `Editor` component? (I'm just curious, this is not a request to make any changes here 🙂)

noisysocks replied:
It doesn't have anything to do with the primary purpose of the hook (building `onNavigateToEntityRecord` callbacks) so I split it out. Feels weird to me that you'd pass `initialPostType` and `initialPostId` to a hook and it returns exactly what you gave it in an object without doing anything else.

andrewserong replied:
Ah, gotcha. That makes sense, thanks!

andrewserong reviewed (approved):
Just gave this a re-test and it's still all working nicely for me in the post and site editors 👍 

The code change is also looking good to me, and seems consistent with the feedback that Riad gave in reviews, too, with relevant code moved over to the template part block where possible.

LGTM! ✨

noisysocks commented:
> > What do you mean by root padding?
> 
> I'll open a separate issue for this.

Thank you! That'd be great.

> In this PR you can see the same thing happen when you double click to edit the template around the page content, but not when you double click to edit the template part. That region previously responded to a double click but doesn't any longer.

Ah gotcha. I'm not sure. Personally I don't really like that dialog at all 😅 I think it would be unexpected if double clicking a template part behaved differently to double clicking a different block. Maybe the designers in this thread can weigh in on what we should do here. I can address it in a follow up if need be.

Mamaduka commented:
@noisysocks, one of the e2e tests became flaky after this got merged into the trunk - https://github.com/WordPress/gutenberg/issues/60102.

jameskoster commented:
> How can we edit the post title on pages in the site editor if the template doesn't include a post title

https://github.com/WordPress/gutenberg/issues/52763

youknowriad commented:
> How can we edit the post title on pages in the site editor if the template doesn't include a post title

You can also disable "template previewing" from the "template panel" in the sidebar and you'll be thrown into the regular "post only" mode that does have a "title" input (like the post editor)

retrofox commented on packages/block-editor/src/components/block-toolbar/index.js:
```diff
@@ -170,26 +170,29 @@ export function PrivateBlockToolbar( {
 					<BlockBindingsIndicator />
 				) }
-				{ ( shouldShowVisualToolbar || isMultiToolbar ) &&
-					isDefaultEditingMode && (
```
Removing this condition makes the block toolbar controls visible regardless of the editing mode. It affects part of the Product Editor UI. [Here is the issue](https://github.com/woocommerce/woocommerce/issues/45908).

Before this change, when the editing mode was contentOnly, the controls that didn't allow editing the block content were not visible.
Below, the custom TextArea field block shows the alignment and regular format buttons.

<img width="691" alt="image" src="https://github.com/WordPress/gutenberg/assets/77539/d459ce0c-975d-4b74-b6d0-15d79d509da5">

After this change, the toolbar shows the parent block, which doesn't do anything since its template is locked to contentOnly:

<img width="481" alt="Screenshot 2024-04-01 at 16 43 25" src="https://github.com/WordPress/gutenberg/assets/77539/c192e716-26fc-4b5e-9229-080d04c36ae0">

Could we consider it a breaking change in terms of the UI?
How could we hide it, recreating the previous state before merging these changes?

noisysocks replied:
Thanks for flagging. I'll open a fix.

noisysocks replied:
I've opened https://github.com/WordPress/gutenberg/pull/60647 which dials this change back to only show the icon on _pattern_ blocks with `contentOnly`.

retrofox replied:
Nice! Thanks a bunch. I see it's already 🍏
